// @react-native-windows/cli/lib-commonjs/e2etest/dependencyConfig.test.js
"use strict";
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 * @format
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const dependencyConfig_1 = require("../config/dependencyConfig");
const generator_common_1 = require("../generator-common");
const templateRoot = path_1.default.resolve('../../../vnext/template');
const testProjectGuid = '{416476D5-974A-4EE2-8145-4E331297247E}';
async function tryMkdir(dir) {
    try {
        await fs_1.default.promises.mkdir(dir, { recursive: true });
    }
    catch (err) { }
}
function project(name, setup) {
    return [name, setup];
}
const projects = [
    // Nothing but a react-native.config.js, with nulls
    project('BlankLib'),
    // Nothing but a windows folder
    project('MissingProjectFilesLib', async (folder) => {
        const windowsDir = path_1.default.join(folder, 'windows');
        await tryMkdir(windowsDir);
    }),
    // New C++ app project based on the template
    project('SimpleCppApp', async (folder) => {
        const windowsDir = path_1.default.join(folder, 'windows');
        await tryMkdir(windowsDir);
        const replacements = {
            name: 'SimpleCppApp',
            namespace: 'SimpleCppApp',
            useMustache: true,
            projectGuidUpper: testProjectGuid,
            projectGuidLower: testProjectGuid.toLowerCase(),
        };
        await generator_common_1.copyAndReplace(path_1.default.join(templateRoot, 'cpp-app/proj/MyApp.sln'), path_1.default.join(windowsDir, 'SimpleCppApp.sln'), replacements, null);
        const projDir = path_1.default.join(windowsDir, 'SimpleCppApp');
        await tryMkdir(projDir);
        await generator_common_1.copyAndReplace(path_1.default.join(templateRoot, 'cpp-app/proj/MyApp.vcxproj'), path_1.default.join(projDir, 'SimpleCppApp.vcxproj'), replacements, null);
    }),
    // New C++ project based on the template
    project('SimpleCppLib', async (folder) => {
        const windowsDir = path_1.default.join(folder, 'windows');
        await tryMkdir(windowsDir);
        const replacements = {
            name: 'SimpleCppLib',
            namespace: 'SimpleCppLib',
            useMustache: true,
            projectGuidUpper: testProjectGuid,
            projectGuidLower: testProjectGuid.toLowerCase(),
        };
        await generator_common_1.copyAndReplace(path_1.default.join(templateRoot, 'cpp-lib/proj/MyLib.sln'), path_1.default.join(windowsDir, 'SimpleCppLib.sln'), replacements, null);
        const projDir = path_1.default.join(windowsDir, 'SimpleCppLib');
        await tryMkdir(projDir);
        await generator_common_1.copyAndReplace(path_1.default.join(templateRoot, 'cpp-lib/proj/MyLib.vcxproj'), path_1.default.join(projDir, 'SimpleCppLib.vcxproj'), replacements, null);
    }),
    // New C# app project based on the template
    project('SimpleCSharpApp', async (folder) => {
        const windowsDir = path_1.default.join(folder, 'windows');
        await tryMkdir(windowsDir);
        const replacements = {
            name: 'SimpleCSharpApp',
            namespace: 'SimpleCSharpApp',
            useMustache: true,
            projectGuidUpper: testProjectGuid,
            projectGuidLower: testProjectGuid.toLowerCase(),
        };
        await generator_common_1.copyAndReplace(path_1.default.join(templateRoot, 'cs-app/proj/MyApp.sln'), path_1.default.join(windowsDir, 'SimpleCSharpApp.sln'), replacements, null);
        const projDir = path_1.default.join(windowsDir, 'SimpleCSharpApp');
        await tryMkdir(projDir);
        await generator_common_1.copyAndReplace(path_1.default.join(templateRoot, 'cs-app/proj/MyApp.csproj'), path_1.default.join(projDir, 'SimpleCSharpApp.csproj'), replacements, null);
    }),
    // New C# project based on the template
    project('SimpleCSharpLib', async (folder) => {
        const windowsDir = path_1.default.join(folder, 'windows');
        await tryMkdir(windowsDir);
        const replacements = {
            name: 'SimpleCSharpLib',
            namespace: 'SimpleCSharpLib',
            useMustache: true,
            projectGuidUpper: testProjectGuid,
            projectGuidLower: testProjectGuid.toLowerCase(),
        };
        await generator_common_1.copyAndReplace(path_1.default.join(templateRoot, 'cs-lib/proj/MyLib.sln'), path_1.default.join(windowsDir, 'SimpleCSharpLib.sln'), replacements, null);
        const projDir = path_1.default.join(windowsDir, 'SimpleCSharpLib');
        await tryMkdir(projDir);
        await generator_common_1.copyAndReplace(path_1.default.join(templateRoot, 'cs-lib/proj/MyLib.csproj'), path_1.default.join(projDir, 'SimpleCSharpLib.csproj'), replacements, null);
    }),
    project('WithIndirectDependency'),
];
// Tests that given userConfig is null, the result will always be null
test.each(projects)('dependencyConfig - %s (userConfig is null)', async (name, setup) => {
    const folder = path_1.default.resolve('src/e2etest/projects/', name);
    if (setup !== undefined) {
        await setup(folder);
    }
    const userConfig = null;
    expect(dependencyConfig_1.dependencyConfigWindows(folder, userConfig)).toBeNull();
});
// Tests the result given a windows project config in react-native.config.js
test.each(projects)('dependencyConfig - %s (Use react-native.config.js)', async (name, setup) => {
    const folder = path_1.default.resolve('src/e2etest/projects/', name);
    const rnc = require(path_1.default.join(folder, 'react-native.config.js'));
    if (setup !== undefined) {
        await setup(folder);
    }
    const userConfig = rnc.dependency.platforms.windows;
    if (name === 'BlankLib') {
        expect(dependencyConfig_1.dependencyConfigWindows(folder, userConfig)).toMatchSnapshot();
    }
    else if (name.endsWith('App')) {
        expect(dependencyConfig_1.dependencyConfigWindows(folder, userConfig)).toMatchSnapshot({
            folder: expect.stringContaining(name),
            projects: expect.arrayContaining([
                expect.objectContaining({
                    projectFile: expect.stringMatching(/Error: .*\.(?:cs|vcx)proj is type '\w+'/),
                }),
            ]),
        });
    }
    else {
        expect(dependencyConfig_1.dependencyConfigWindows(folder, userConfig)).toMatchSnapshot({
            folder: expect.stringContaining(name),
        });
    }
});
// Tests the result of ignoring the windows project config in react-native.config.js
test.each(projects)('dependencyConfig - %s (Ignore react-native.config.js)', async (name, setup) => {
    const folder = path_1.default.resolve('src/e2etest/projects/', name);
    if (setup !== undefined) {
        await setup(folder);
    }
    const userConfig = {};
    if (name === 'BlankLib') {
        expect(dependencyConfig_1.dependencyConfigWindows(folder, userConfig)).toMatchSnapshot();
    }
    else if (name.endsWith('App')) {
        expect(dependencyConfig_1.dependencyConfigWindows(folder, userConfig)).toMatchSnapshot({
            folder: expect.stringContaining(name),
            projects: expect.arrayContaining([
                expect.objectContaining({
                    projectFile: expect.stringMatching(/Error: .*\.(?:cs|vcx)proj is type '\w+'/),
                }),
            ]),
        });
    }
    else {
        expect(dependencyConfig_1.dependencyConfigWindows(folder, userConfig)).toMatchSnapshot({
            folder: expect.stringContaining(name),
        });
    }
});
