// @react-native-windows/cli/lib-commonjs/e2etest/projectConfig.utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ensureWinUI3Project = exports.tryMkdir = exports.testProjectGuid = exports.templateRoot = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const generator_common_1 = require("../generator-common");
exports.templateRoot = path_1.default.resolve('../../../vnext/template');
exports.testProjectGuid = '{416476D5-974A-4EE2-8145-4E331297247E}';
async function tryMkdir(dir) {
    try {
        await fs_1.default.promises.mkdir(dir, { recursive: true });
    }
    catch (err) { }
}
exports.tryMkdir = tryMkdir;
async function ensureWinUI3Project(folder) {
    const windowsDir = path_1.default.join(folder, 'windows');
    if (fs_1.default.existsSync(windowsDir)) {
        await fs_1.default.promises.rmdir(windowsDir, { recursive: true });
    }
    await tryMkdir(windowsDir);
    const replacements = {
        name: 'WithWinUI3',
        namespace: 'WithWinUI3',
        useMustache: true,
        projectGuidUpper: exports.testProjectGuid,
        projectGuidLower: exports.testProjectGuid.toLowerCase(),
        useWinUI3: false,
        useHermes: false,
        useExperimentalNuget: false,
        packagesConfigCppNugetPackages: [
            {
                id: 'Microsoft.ReactNative.Cxx',
                version: '1.0.0',
                hasProps: false,
                hasTargets: true,
            },
            {
                id: 'Microsoft.UI.Xaml',
                version: '2.3.4.5',
                hasProps: false,
                hasTargets: false,
            },
            {
                id: 'Microsoft.WinUI',
                version: '3.2.1.0',
                hasProps: false,
                hasTargets: false,
            },
        ],
    };
    await generator_common_1.copyAndReplace(path_1.default.join(exports.templateRoot, 'cpp-app/proj/MyApp.sln'), path_1.default.join(windowsDir, 'WithWinUI3.sln'), replacements, null);
    const projDir = path_1.default.join(windowsDir, 'WithWinUI3');
    await tryMkdir(projDir);
    await generator_common_1.copyAndReplace(path_1.default.join(exports.templateRoot, 'cpp-app/proj/MyApp.vcxproj'), path_1.default.join(projDir, 'WithWinUI3.vcxproj'), replacements, null);
    await generator_common_1.copyAndReplace(path_1.default.join(exports.templateRoot, 'cpp-app/proj/packages.config'), path_1.default.join(projDir, 'packages.config'), replacements, null);
    await generator_common_1.copyAndReplace(path_1.default.join(exports.templateRoot, 'shared-app/proj/ExperimentalFeatures.props'), path_1.default.join(windowsDir, 'ExperimentalFeatures.props'), replacements, null);
}
exports.ensureWinUI3Project = ensureWinUI3Project;
